expan/core/experiment.py

```diff
@@ -290,20 +294,51 @@ def _quantile_filtering(self, data, kpis, percentile, threshold_type):
         
         :param kpis: the kpis to perform filtering
         :type  kpis: list[str]
-        :param percentile: percentile considered as filtering threshold
-        :type  percentile: float
-        :param threshold_type: type of threshold used ('lower' or 'upper')
-        :type  threshold_type: str
+        :param thresholds: dict of thresholds mapping KPI names to (type, percentile) tuples
+        :type  thresholds: dict
 
         :return: boolean values indicating whether the row should be filtered
         :rtype: pd.Series
         """
-        method_table   = {'upper': lambda x: x > threshold, 'lower': lambda x: x <= threshold}
-        na_replacement = {'upper': -float_info.max,         'lower':  float_info.max}
 
-        for column in data[kpis].columns:
-            threshold = np.percentile(data[column].fillna(na_replacement[threshold_type]), percentile)
-            flags = flags | data[column].apply(method_table[threshold_type])
+        def find_smallest(data, quantile):
+            """ Return indices of data points smaller than given quantile."""
+            threshold = data.quantile(quantile)
+            return data.apply(lambda x: x < threshold)
+
+        def find_largest(data, quantile):
+            """ Return indices of data points larger than given quantile."""
+            threshold = data.quantile(quantile)
+            return data.apply(lambda x: x > threshold)
+
+        def find_smallest_and_largest(data, quantile):
+            """ Return indices of data points outside of the given quantile."""
+            rest = 1.0 - quantile
+            quantiles = [rest/2.0, 1.0 - rest/2.0]
+            thresholds = list(data.quantile(quantiles))
+            return data.apply(lambda x: x < thresholds[0] or x > thresholds[1])
+
+        method_table = {'upper': find_largest,
+                        'lower': find_smallest,
+                        'two-sided': find_smallest_and_largest}
+
+        for col in data[kpis].columns:
+            column = data[col].copy()
+            column.replace([np.inf, -np.inf], np.nan)
+
+            quantile = 0.99
+            threshold_type = _choose_threshold_type(column)
```

Comment:
reassignment below is a bit confusing, maybe this can go to an `else` branch. 
Though, super minor